include/pushmi/time_source.h

```diff
@@ -0,0 +1,481 @@
+public:
+  using properties = property_set<is_time<>, is_executor<>, is_single<>>;
+
+  time_source_executor(
+    std::shared_ptr<time_source_shared<E, time_point>> source,
+    std::shared_ptr<time_source_queue<E, time_point, NF, Executor>> queue) :
+    source_(std::move(source)), queue_(std::move(queue)) {}
+
+  auto top() { return queue_->nf_(); }
+  auto executor() { return *this; }
+
+  PUSHMI_TEMPLATE(class TPA, class Out)
+    (requires Regular<TPA> && Receiver<Out, is_single<>>)
+  void submit(TPA tp, Out out) {
+    // queue for later
+    source_->insert(queue_, time_heap_item<E, TP>{tp, any_single<any_time_executor_ref<E, TP>, E>{std::move(out)}});
+  }
+};
+
+//
+// the time executor factory produces a new time ordered queue each time that it is called.
+//
+
+template<class E, class TP, class NF, class ExecutorFactory>
+class time_source_executor_factory_fn {
+  using time_point = std::decay_t<TP>;
+  std::shared_ptr<time_source_shared<E, time_point>> source_;
+  NF nf_;
+  ExecutorFactory ef_;
+public:
+  time_source_executor_factory_fn(
+    std::shared_ptr<time_source_shared<E, time_point>> source,
+    NF nf,
+    ExecutorFactory ef
+  ) : source_(std::move(source)), nf_(std::move(nf)), ef_(std::move(ef)) {}
+  auto operator()(){
+    auto ex = ef_();
+    auto queue = std::make_shared<time_source_queue<E, time_point, NF, decltype(ex)>>(
+      source_,
+      nf_,
+      std::move(ex));
+    return time_source_executor<E, time_point, NF, decltype(ex)>{source_, queue};
+  }
+};
+
+//
+// each time_source is an independent source of timed events
+//
+// a time_source is a time_single_executor factory, it is not an executor itself.
+//
+// each time_source has a single thread that is shared across all the
+// time executors it produces. the thread is used to wait for the next time event.
+// when a time event is ready the thread will use the executor passed into make()
+// to callback on the receiver passed to the time executor submit()
+//
+// passing an executor to time_source.make() will create a time executor factory.
+// the time executor factory is a function that will return a time executor when
+// called with no arguments.
+//
+//
+//
+
+template<class E = std::exception_ptr, class TP = std::chrono::system_clock::time_point>
+class time_source {
+public:
+  using time_point = std::decay_t<TP>;
+private:
+  std::shared_ptr<time_source_shared<E, time_point>> source_;
+public:
+  time_source() : source_(std::make_shared<time_source_shared<E, time_point>>()) {
+    source_->start(source_);
+  };
+
+  PUSHMI_TEMPLATE(class NF, class ExecutorFactory)
+    (requires Invocable<ExecutorFactory&> && Executor<invoke_result_t<ExecutorFactory&>>)
+  auto make(NF nf, ExecutorFactory ef) {
+    return time_source_executor_factory_fn<E, time_point, NF, ExecutorFactory>{source_, std::move(nf), std::move(ef)};
+  }
+
+  void join() {
+    source_->join(source_);
+  }
+};
```

Comment:
Does this type have value semantics?

Reply:
yes. start and join are static functions that take a shared_ptr so that they do not have to use shared_from_this.